### ISSUE TITLE
MediaDrmCallback methods throwing MediaDrmCallbackException in ExoPlayer 2.12.0+

### ISSUE THREAD

**Issue by qvk** [google/ExoPlayer#8683]
### Use case description
`MediaDrmCallback` methods throw `MediaDrmCallbackException` in ExoPlayer 2.12.0+ since [this commit](https://github.com/google/ExoPlayer/commit/025a2c2b6240b7391ec8965c33ea066ed8b9563d#diff-0e43a325667d07e99c8aed34c9f786f98cf656818034c5d6fc90f4b3fe955f25).

Prior to ExoPlayer 2.12.0, we had a custom implementation of `MediaDrmCallback` interface which throws our own custom exceptions (and other not our own checked exceptions, e.g. `KeysExpiredException`) which could be interpreted appropriately by implementing `onPlayerError`. However, now that `MediaDrmCallback` methods throw a `MediaDrmCallbackException` which is a checked exception, that is not straightforward anymore. 

`MediaDrmCallbackException` (unlike `IOException`, etc.) accepts very specific parameters `(dataSpec, uriAfterRedirects, etc.)` which may not be available or relevant  in a custom implementation of `MediaDrmCallback`.

We could
(1) throw `MediaDrmCallbackException(null, null, null, null, ourIntendedException /* cause */)`

(2) make our custom exceptions runtime exceptions, however this still won't cleanly allow throwing checked exceptions not in our control (e.g. `KeysExpiredException`)

### Proposed solution
Does it make sense to revert back to the older interface definition, or make `MediaDrmCallbackException` runtime exception? 

What should be the ideal way to handle a custom `MediaDrmCallback` implementation which doesn't have the relevant information for `(dataSpec, uriAfterRedirects, responseHeaders, bytesLoaded)`?

Thanks

**Comment by AquilesCanta** [comment]
You can still throw runtime exceptions. They just won't get retried. I would think that retrial only makes sense for IO Exceptions. So I think we need to understand better your usecase to provide a response.

(1) Yes. I wouldn't use null, but rather empty parameters.

For (2), there's not much context so I will make some assumptions. MediaDrmCallback is intended to perform a key request, so KeysExpiredException should not be thrown from it. I'm guessing you are doing some form of key caching (which can expire?) and then either you don't want to retry (in which case you don't need MediaDrmCallbackException), or you want to retry ignoring the "cached keys", in which case you can wrap in a MediaDrmCallbackException, or you can just retry yourself without the DrmSession learning about it (which seems to make more sense anyway).

In conclusion, please provide more context on how your usecase is specifically affected and we can take a look.

**Comment by qvk** [comment]
Regarding (2), yes we are doing something like that in one of our use cases involving cached keys, where we don't want to retry on key expiry.

For non-cached keys use case (where the license request has to be done every time a playback is requested), our license proxy server bundles additional business specific data along with the license response (license_response_blob + custom_data). From the response, license_response_blob is returned from executeKeyRequest as expected, while custom_data may contain additionally either some_result or error information from proxy server. A custom exception can be thrown depending on the contents of the error description (detailed response beyond a 403/404 which is already covered by the default HttpMediaDrmCallback in the library) received from proxy server (related to authentication, etc.), or simply a JSONException, etc. while parsing the response. However, these exceptions will need to be wrapped in MediaDrmCallbackException now.

These are our use cases. Does it warrant any change, or going with (1) is the way forward?

Thanks :)

**Comment by AquilesCanta** [comment]
Anything that does not need to be retried, like described in 

> involving cached keys, where we don't want to retry on key expiry.

can be thrown as it is. The [won't be retried](https://github.com/google/ExoPlayer/blob/4364b9150f236813a2070fc79d3bcb76e11969ac/library/core/src/main/java/com/google/android/exoplayer2/drm/DefaultDrmSession.java#L613).

I'm assuming the following:

> A custom exception can be thrown depending on the contents of the error description

Might want to be retried. In this case, you should  wrap them in in a MediaDrmCallbackException (from what you describe, you might have most required info, like loaded bytes, URI, etc). You can then customize the way it's retried by injecting a custom LoadErrorHandlingPolicy which potentially extends the default one. Is there something fundamentally problematic with wrapping in a MediaDrmCallbackException?

I appreciate the API change is annoying but from my point of view the worst case scenario is you having to wrap your previous `executeKeyRequest` like so:

```
 @Override
  public byte[] executeKeyRequest(UUID uuid, KeyRequest request) throws MediaDrmCallbackException {
    try {
      return oldExecuteKeyRequest(...);
    catch (Exception e) {
      throw new MediaDrmCallbackException(..., cause);
    }
 }
```

But the benefit you get is that you can use a custom load error handling policy to customize the retry behavior to better suit your needs (e.g. by not retrying things you know you don't need to retry, which was not possible before). Let me know if this answers your question.

**Comment by qvk** [comment]
Thanks for the clarifications. I'll work on acquiring more relevant parameters in our implementation required for MediaDrmCallbackException parameters, so it can be used to wrap our exceptions. I'll also tweak our code to make appropriate use of the custom load error handling policy.

Yes, I think I have everything to go forward with this. Thanks!
